### PR TITLE
[Fixed label cluster]: Implement fixed label cluster server callback

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
What is being fixed?  Examples:
* Missing implementation for  fixed label cluster in CHIP sdk

#### Change overview
Implement fixed label cluster server callback

#### Testing
How was this tested? (at least one bullet point required)
```
yufengw@yufengw-SEi:~/connectedhomeip/out/debug/standalone$ ./chip-tool fixedlabel read label-list 12344321 0
....
[1639067914.490591][602487:602492] CHIP:DMG: ReportDataMessage =
[1639067914.490597][602487:602492] CHIP:DMG: {
[1639067914.490601][602487:602492] CHIP:DMG: 	AttributeReportIBs =
[1639067914.490610][602487:602492] CHIP:DMG: 	[
[1639067914.490614][602487:602492] CHIP:DMG: 		AttributeReportIB =
[1639067914.490623][602487:602492] CHIP:DMG: 		{
[1639067914.490629][602487:602492] CHIP:DMG: 			AttributeDataIB =
[1639067914.490637][602487:602492] CHIP:DMG: 			{
[1639067914.490644][602487:602492] CHIP:DMG: 				DataVersion = 0x0,
[1639067914.490651][602487:602492] CHIP:DMG: 				AttributePathIB =
[1639067914.490658][602487:602492] CHIP:DMG: 				{
[1639067914.490666][602487:602492] CHIP:DMG: 					Endpoint = 0x0,
[1639067914.490674][602487:602492] CHIP:DMG: 					Cluster = 0x40,
[1639067914.490681][602487:602492] CHIP:DMG: 					Attribute = 0x0000_0000,
[1639067914.490688][602487:602492] CHIP:DMG: 				}
[1639067914.490696][602487:602492] CHIP:DMG: 					
[1639067914.490704][602487:602492] CHIP:DMG: 					Data = [
[1639067914.490711][602487:602492] CHIP:DMG: 						
[1639067914.490719][602487:602492] CHIP:DMG: 					],
[1639067914.490725][602487:602492] CHIP:DMG: 			},
[1639067914.490733][602487:602492] CHIP:DMG: 			
[1639067914.490738][602487:602492] CHIP:DMG: 		},
[1639067914.490749][602487:602492] CHIP:DMG: 		
[1639067914.490755][602487:602492] CHIP:DMG: 		AttributeReportIB =
[1639067914.490764][602487:602492] CHIP:DMG: 		{
[1639067914.490769][602487:602492] CHIP:DMG: 			AttributeDataIB =
[1639067914.490775][602487:602492] CHIP:DMG: 			{
[1639067914.490782][602487:602492] CHIP:DMG: 				DataVersion = 0x0,
[1639067914.490788][602487:602492] CHIP:DMG: 				AttributePathIB =
[1639067914.490795][602487:602492] CHIP:DMG: 				{
[1639067914.490803][602487:602492] CHIP:DMG: 					Endpoint = 0x0,
[1639067914.490811][602487:602492] CHIP:DMG: 					Cluster = 0x40,
[1639067914.490818][602487:602492] CHIP:DMG: 					Attribute = 0x0000_0000,
[1639067914.490825][602487:602492] CHIP:DMG: 					ListIndex = Null,
[1639067914.490832][602487:602492] CHIP:DMG: 				}
[1639067914.490841][602487:602492] CHIP:DMG: 					
[1639067914.490847][602487:602492] CHIP:DMG: 					Data = 
[1639067914.490855][602487:602492] CHIP:DMG: 					{
[1639067914.490863][602487:602492] CHIP:DMG: 						0x1 = "room", 
[1639067914.490872][602487:602492] CHIP:DMG: 						0x2 = "bedroom 2", 
[1639067914.490880][602487:602492] CHIP:DMG: 					},
[1639067914.490886][602487:602492] CHIP:DMG: 			},
[1639067914.490895][602487:602492] CHIP:DMG: 			
[1639067914.490900][602487:602492] CHIP:DMG: 		},
[1639067914.490912][602487:602492] CHIP:DMG: 		
[1639067914.490917][602487:602492] CHIP:DMG: 		AttributeReportIB =
[1639067914.490926][602487:602492] CHIP:DMG: 		{
[1639067914.490933][602487:602492] CHIP:DMG: 			AttributeDataIB =
[1639067914.490941][602487:602492] CHIP:DMG: 			{
[1639067914.490948][602487:602492] CHIP:DMG: 				DataVersion = 0x0,
[1639067914.490954][602487:602492] CHIP:DMG: 				AttributePathIB =
[1639067914.490960][602487:602492] CHIP:DMG: 				{
[1639067914.490966][602487:602492] CHIP:DMG: 					Endpoint = 0x0,
[1639067914.490973][602487:602492] CHIP:DMG: 					Cluster = 0x40,
[1639067914.490981][602487:602492] CHIP:DMG: 					Attribute = 0x0000_0000,
[1639067914.490988][602487:602492] CHIP:DMG: 					ListIndex = Null,
[1639067914.490996][602487:602492] CHIP:DMG: 				}
[1639067914.491004][602487:602492] CHIP:DMG: 					
[1639067914.491011][602487:602492] CHIP:DMG: 					Data = 
[1639067914.491019][602487:602492] CHIP:DMG: 					{
[1639067914.491026][602487:602492] CHIP:DMG: 						0x1 = "orientation", 
[1639067914.491034][602487:602492] CHIP:DMG: 						0x2 = "North", 
[1639067914.491041][602487:602492] CHIP:DMG: 					},
[1639067914.491048][602487:602492] CHIP:DMG: 			},
[1639067914.491057][602487:602492] CHIP:DMG: 			
[1639067914.491062][602487:602492] CHIP:DMG: 		},
[1639067914.491074][602487:602492] CHIP:DMG: 		
[1639067914.491079][602487:602492] CHIP:DMG: 		AttributeReportIB =
[1639067914.491087][602487:602492] CHIP:DMG: 		{
[1639067914.491093][602487:602492] CHIP:DMG: 			AttributeDataIB =
[1639067914.491099][602487:602492] CHIP:DMG: 			{
[1639067914.491106][602487:602492] CHIP:DMG: 				DataVersion = 0x0,
[1639067914.491113][602487:602492] CHIP:DMG: 				AttributePathIB =
[1639067914.491119][602487:602492] CHIP:DMG: 				{
[1639067914.491126][602487:602492] CHIP:DMG: 					Endpoint = 0x0,
[1639067914.491134][602487:602492] CHIP:DMG: 					Cluster = 0x40,
[1639067914.491142][602487:602492] CHIP:DMG: 					Attribute = 0x0000_0000,
[1639067914.491149][602487:602492] CHIP:DMG: 					ListIndex = Null,
[1639067914.491156][602487:602492] CHIP:DMG: 				}
[1639067914.491163][602487:602492] CHIP:DMG: 					
[1639067914.491170][602487:602492] CHIP:DMG: 					Data = 
[1639067914.491177][602487:602492] CHIP:DMG: 					{
[1639067914.491185][602487:602492] CHIP:DMG: 						0x1 = "floor", 
[1639067914.491193][602487:602492] CHIP:DMG: 						0x2 = "2", 
[1639067914.491201][602487:602492] CHIP:DMG: 					},
[1639067914.491208][602487:602492] CHIP:DMG: 			},
[1639067914.491216][602487:602492] CHIP:DMG: 			
[1639067914.491221][602487:602492] CHIP:DMG: 		},
[1639067914.491233][602487:602492] CHIP:DMG: 		
[1639067914.491238][602487:602492] CHIP:DMG: 		AttributeReportIB =
[1639067914.491247][602487:602492] CHIP:DMG: 		{
[1639067914.491252][602487:602492] CHIP:DMG: 			AttributeDataIB =
[1639067914.491259][602487:602492] CHIP:DMG: 			{
[1639067914.491265][602487:602492] CHIP:DMG: 				DataVersion = 0x0,
[1639067914.491272][602487:602492] CHIP:DMG: 				AttributePathIB =
[1639067914.491278][602487:602492] CHIP:DMG: 				{
[1639067914.491285][602487:602492] CHIP:DMG: 					Endpoint = 0x0,
[1639067914.491293][602487:602492] CHIP:DMG: 					Cluster = 0x40,
[1639067914.491300][602487:602492] CHIP:DMG: 					Attribute = 0x0000_0000,
[1639067914.491308][602487:602492] CHIP:DMG: 					ListIndex = Null,
[1639067914.491315][602487:602492] CHIP:DMG: 				}
[1639067914.491323][602487:602492] CHIP:DMG: 					
[1639067914.491329][602487:602492] CHIP:DMG: 					Data = 
[1639067914.491337][602487:602492] CHIP:DMG: 					{
[1639067914.491344][602487:602492] CHIP:DMG: 						0x1 = "direction", 
[1639067914.491353][602487:602492] CHIP:DMG: 						0x2 = "up", 
[1639067914.491360][602487:602492] CHIP:DMG: 					},
[1639067914.491367][602487:602492] CHIP:DMG: 			},
[1639067914.491375][602487:602492] CHIP:DMG: 			
[1639067914.491380][602487:602492] CHIP:DMG: 		},
[1639067914.491388][602487:602492] CHIP:DMG: 		
[1639067914.491394][602487:602492] CHIP:DMG: 	],
[1639067914.491413][602487:602492] CHIP:DMG: 	
[1639067914.491417][602487:602492] CHIP:DMG: 	SuppressResponse = true, 
[1639067914.491422][602487:602492] CHIP:DMG: }
[1639067914.491581][602487:602492] CHIP:TOO: FixedLabel.LabelList response: 4 entries
[1639067914.491602][602487:602492] CHIP:TOO:   [1]: {
[1639067914.491607][602487:602492] CHIP:TOO:     Label: room
[1639067914.491612][602487:602492] CHIP:TOO:     Value: bedroom 2
[1639067914.491617][602487:602492] CHIP:TOO:   }
[1639067914.491625][602487:602492] CHIP:TOO:   [2]: {
[1639067914.491629][602487:602492] CHIP:TOO:     Label: orientation
[1639067914.491634][602487:602492] CHIP:TOO:     Value: North
[1639067914.491638][602487:602492] CHIP:TOO:   }
[1639067914.491646][602487:602492] CHIP:TOO:   [3]: {
[1639067914.491651][602487:602492] CHIP:TOO:     Label: floor
[1639067914.491655][602487:602492] CHIP:TOO:     Value: 2
[1639067914.491660][602487:602492] CHIP:TOO:   }
[1639067914.491666][602487:602492] CHIP:TOO:   [4]: {
[1639067914.491671][602487:602492] CHIP:TOO:     Label: direction
[1639067914.491676][602487:602492] CHIP:TOO:     Value: up
[1639067914.491681][602487:602492] CHIP:TOO:   }
[1639067914.491701][602487:602492] CHIP:DMG: Client[0] moving to [UNINIT]
```

### CODE REVIEW

File: src/app/clusters/fixed-label-server/fixed-label-server.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+CHIP_ERROR DescriptorAttrAccess::ReadLabelList(EndpointId endpoint, AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    DeviceLayer::LabelList<DeviceLayer::kMaxFixedLabels> labelList;
+
+    if (DeviceLayer::PlatformMgr().GetFixedLabelList(labelList) == CHIP_NO_ERROR)
+    {
+        err = aEncoder.EncodeList([&labelList](const auto & encoder) -> CHIP_ERROR {
+            for (auto label : labelList)
```
Comment (bzbarsky-apple) [inline src/app/clusters/fixed-label-server/fixed-label-server.cpp:66]:
```suggestion
            for (auto & label : labelList)
```

File: src/app/clusters/fixed-label-server/fixed-label-server.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            {
+                ReturnErrorOnFailure(encoder.Encode(label));
+            }
+
+            return CHIP_NO_ERROR;
+        });
+    }
+    else
+    {
+        err = aEncoder.Encode(DataModel::List<Structs::LabelStruct::Type>());
```
Comment (bzbarsky-apple) [inline src/app/clusters/fixed-label-server/fixed-label-server.cpp:76]: I wonder whether we should have an EncodeEmptyList helper... followup for that.

File: src/app/clusters/fixed-label-server/fixed-label-server.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::FixedLabel;
+using namespace chip::app::Clusters::FixedLabel::Attributes;
+
+namespace {
+
+class DescriptorAttrAccess : public AttributeAccessInterface
```
Comment (bzbarsky-apple) [inline src/app/clusters/fixed-label-server/fixed-label-server.cpp:41]: This is not the right name.  Copy/paste problem?

File: src/app/clusters/fixed-label-server/fixed-label-server.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    }
+    else
+    {
+        err = aEncoder.Encode(DataModel::List<Structs::LabelStruct::Type>());
+    }
+
+    return err;
+}
+
+CHIP_ERROR DescriptorAttrAccess::ReadClusterRevision(EndpointId endpoint, AttributeValueEncoder & aEncoder)
```
Comment (bzbarsky-apple) [inline src/app/clusters/fixed-label-server/fixed-label-server.cpp:82]: Why is this not coming from the ZAP configuration?

File: src/app/clusters/fixed-label-server/fixed-label-server.cpp [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+};
+
+constexpr uint16_t DescriptorAttrAccess::ClusterRevision;
+
+CHIP_ERROR DescriptorAttrAccess::ReadLabelList(EndpointId endpoint, AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    DeviceLayer::LabelList<DeviceLayer::kMaxFixedLabels> labelList;
+
+    if (DeviceLayer::PlatformMgr().GetFixedLabelList(labelList) == CHIP_NO_ERROR)
```
Comment (bzbarsky-apple) [inline src/app/clusters/fixed-label-server/fixed-label-server.cpp:63]: This doesn't make sense to me.  Shouldn't the list depend on the endpoint?

File: src/include/platform/PlatformManager.h [hunk 6]
```diff
@@ -396,5 +399,10 @@ inline CHIP_ERROR PlatformManager::StartChipTimer(System::Clock::Timeout duratio
     return static_cast<ImplClass *>(this)->_StartChipTimer(duration);
 }
 
+inline CHIP_ERROR PlatformManager::GetFixedLabelList(LabelList<kMaxFixedLabels> & labelList)
```
Comment (bzbarsky-apple) [inline src/include/platform/PlatformManager.h:402]: This API does not make sense, given labels are per-endpoint.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
The main problem is that the basic setup with the configuration manager in this PR does not make sense.

File: src/app/clusters/fixed-label-server.cpp [hunk 7]
```diff
@@ -0,0 +1,101 @@
+/*
```
Comment (bzbarsky-apple) [inline src/app/clusters/fixed-label-server.cpp:1]: This file should not be here?

File: src/platform/Linux/PlatformManagerImpl.cpp [hunk 8]
```diff
@@ -261,6 +261,34 @@ CHIP_ERROR PlatformManagerImpl::_Shutdown()
     return Internal::GenericPlatformManagerImpl_POSIX<PlatformManagerImpl>::_Shutdown();
 }
 
+CHIP_ERROR PlatformManagerImpl::_GetFixedLabelList(EndpointId endpoint, LabelList<kMaxFixedLabels> & labelList)
+{
+    // In Linux simulation, return following hardcoded lableList on all endpoints.
```
Comment (bzbarsky-apple) [inline src/platform/Linux/PlatformManagerImpl.cpp:266]:
```suggestion
    // In Linux simulation, return following hardcoded labelList on all endpoints.
```

File: src/app/clusters/fixed-label-server/fixed-label-server.cpp [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+{
+    VerifyOrDie(aPath.mClusterId == FixedLabel::Id);
+
+    switch (aPath.mAttributeId)
+    {
+    case LabelList::Id: {
+        return ReadLabelList(aPath.mEndpointId, aEncoder);
+    }
+    default: {
+        break;
```
Comment (electrocucaracha) [inline src/app/clusters/fixed-label-server/fixed-label-server.cpp:89]: maybe a dumb question but does this default case makes sense?

Comment (yufengwangca) [reply src/app/clusters/fixed-label-server/fixed-label-server.cpp:89]: This 'default' is optional, saying a switch must always contain a default is like saying every 'if-elseif' must contain a 'else'. This is just align the style with other server callback implementations.

File: src/app/clusters/fixed-label-server/fixed-label-server.cpp (hunk 1 above)
Comment (yufengwangca) [reply src/app/clusters/fixed-label-server/fixed-label-server.cpp:61]: use "auto &" or "const auto &" always give following error on Darwin, but not on other platforms

`../../src/app/clusters/fixed-label-server/fixed-label-server.cpp:61:13: error: cannot bind non-const lvalue reference of type ‘chip::app::Clusters::FixedLabel::Structs::LabelStruct::Type&’ to an rvalue of type ‘chip::app::Clusters::FixedLabel::Structs::LabelStruct::Type’`

**Review by yunhanw-google — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**